pknotsRG-1.3/man/pknotsRG.1
.\" Automatically generated by Pod::Man v1.34, Pod::Parser v1.13
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` <
.    ds C' >
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PKNOTSRG 1"
.TH PKNOTSRG 1 "09/19/06" "pknotsRG 1.3" "pknotsRG documentation"
.SH "NAME"
pknotsRG 1.3
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
pknotsRG: Folding canonical \s-1RNA\s0 secondary structures including pseudoknots
.SH "INTRODUCTION"
.IX Header "INTRODUCTION"
pknotsRG is a tool for thermodynamic folding of \s-1RNA\s0 secondary
structures, including the class of canonical simple recursive
pseudoknots. This class and the algorithms are described in detail in:
.PP
\&\*(L"Design, Implementation and Evaluation of a Practical Pseudoknot
Folding Algorithm based on Thermodynamics\*(R", Jens Reeder and Robert
Giegerich, \s-1BMC\s0 Bioinformatics, 5:104, 2004
.PP
Output:
.PP
The output consist of the input sequence, the secondary structure, and
the minimum free energy.  Basepairs of the first pseudoknot helix are
denoted by '{' and '}', the second helix by '[' and ']'. Basepairs not
involved in a pseudoknot correspond to normal brackets, '(' and ')'.
.PP
Example:
.PP
.Vb 2
\&    UCAAGUAUUCCGAAGCUCAACGGGAAAAUGAGCUA
\&    .......[[[[[.{{{{{{.]]]]]...}}}}}}. ( -14.9)
.Ve
.PP
Since pknotsRG-loc (option \-l) finds the best local pseudoknot, the
output of this variant also contains the start and end position of the
corresponding subsequence:
.PP
.Vb 3
\&    7........................34
\&    UUCCGAAGCUCAACGGGAAAAUGAGCU
\&    [[[[[.{{{{{{.]]]]]...}}}}}} (-14.51)
.Ve
.PP
When option \-s is used, pknotsRG computes a list of unordered
suboptimal results.
.PP
The energy model we use for pseudoknots:
.PP
Destabilizing:
.PP
.Vb 3
\&    creating a new pseudoknot :             9.0   (pkinit)
\&    basepair inside pseudoknot:             0.0
\&    not paired base in pk     :             0.3   (npp)
.Ve
.PP
The parameters pkinit and npp have major influence on the pseudoknot
energy. The values shown above seem to be the best choice for our test
sets. However, feel free to experiment with these values.
.PP
Stabilizing:
.PP
.Vb 3
\&    stacking of basepairs     :             stack
\&    base dangling of a pk pair:             dangle
\&    coaxial stacking          :             stack
.Ve
.PP
all values in (kcal/mole)
.PP
stack and dangle are the normal energies for nested structures like in
mfold\-3.1 or RNAfold. If nested or unnested structures occur inside a
pseudoknot, their energy, of course, contributes to the overall
pseudoknot energy.
.SH "OPTIONS"
.IX Header "OPTIONS"
\&\fB\-h\fR            	Display this information
.PP
Option \-h displays a command option overview.
.PP
\&\fB\-H\fR \f(CW\*(C`option\*(C'\fR  	Display detailed information on <option>
.PP
This displays the corresponding section of the pknotsRG manual for
the given command line option.
.PP
\&\fB\-v\fR            	Show version
.PP
This shows the version number of pknotsRG.
.PP
\&\fB\-m\fR                   Use mfe strategy
.PP
Compute the best structure (the structure with minimum free
energy). This is the default strategy.
.PP
\&\fB\-f\fR                   Use enf strategy
.PP
Compute the best structure that actually contains at least one
pseudoknot.
.PP
\&\fB\-l\fR                   Use loc strategy
.PP
Compute the best \*(L"compact\*(R" pseudoknot, defined as the structure with
the lowest energy to length ratio.
.PP
\&\fB\-s\fR                   Show suboptimals
.PP
Show suboptimals. If neither \fB\-e\fR nor \fB\-c\fR is specified, default is
use with \fB\-c\fR 10. This option can be combined with \fB\-m\fR, \fB\-f\fR and
\&\fB\-l\fR.
.PP
\&\fB\-u\fR                   No dangling bases (implies \-s)
.PP
This calculates all structures without dangling bases. Setting this
option reduces the number of suboptimals drastically.
.PP
\&\fB\-o\fR                   No suboptimals inside pknots (implies \-s \-l)
.PP
Used with loc\-strategy, this option discards all suboptimal structures
inside of pseudoknots.
.PP
\&\fB\-e\fR \f(CW\*(C`value\*(C'\fR   	Set energy range (kcal/mol)
.PP
This sets the energy range for suboptimal results. value is the
difference to the minimum free energy for the sequence.
.PP
\&\fB\-c\fR \f(CW\*(C`value\*(C'\fR   	Set energy range (%%)
.PP
This sets the energy range as percentage value of the minimum free
energy. For example, when \-c 10 is specified, and the minimum free
energy is \-10.0 kcal/mol, the energy range is set to \-9.0 to \-10.0
kcal/mol.
.PP
\&\fB\-n\fR \f(CW\*(C`value\*(C'\fR          Set npp-value [0.3]
.PP
This option sets the energy penalty for unpaired bases inside of
pseudoknots.
.PP
\&\fB\-p\fR \f(CW\*(C`value\*(C'\fR          Set pkinit-value [9]
.PP
This options sets the energy penalty for creating a new pseudoknot.
.PP
\&\fB\-k\fR \f(CW\*(C`value\*(C'\fR          Set maximal pknot-length 
.PP
This option sets the maximal pseudoknot-length allowed.
.PP
\&\fB\-w\fR \f(CW\*(C`value\*(C'\fR   	Set window size
.PP
Beginning with position 1 of the input sequence, the analysis is
repeatedly processed on subsequences of the specified size. After each
calculation, the results are printed out and the window is moved by
the window position increment (\-W), until the end of the input
sequence is reached.
.PP
\&\fB\-W\fR \f(CW\*(C`value\*(C'\fR   	Set window position increment
.PP
This option specifies the increment for the window analysis mode
(\-w).
.PP
\&\fB\-S\fR \f(CW\*(C`value\*(C'\fR   	Specify output width for structures
.PP
This splits the structure strings into parts of the specified
length. This option is useful when displaying results for long
sequences that would otherwise not fit onto the screen.
.PP
\&\fB\-F\fR \f(CW\*(C`filename\*(C'\fR	Read input from file
.PP
Let pknotsRG load its input data from file. file can contain a
plain single sequence, or multiple sequences in fasta format. When
given multiple sequences, each sequence is processed separately in the
order of input.
.PP
Valid characters in an input sequence are \*(L"\s-1ACGU\s0\*(R" and \*(L"acgu\*(R". \*(L"T\*(R" and
\&\*(L"t\*(R" will be converted to \*(L"U\*(R". Other letters are mapped to \*(L"N\*(R" and will
not be paired. All other characters are ignored.
.PP
\&\fB\-z\fR            	Colored output
.PP
This option enables colored output. In interactive mode, this is the
default setting, so use \-z to disable colors here.
.Sh "Additional interactive mode commands"
.IX Subsection "Additional interactive mode commands"
\&\fB:s\fR          Show current configuration
.PP
This command shows the current settings in an interactive pknotsRG
session.
.PP
\&\fB:d\fR           Reset configuration
.PP
This command sets all settings to their default values.
.PP
\&\fB:e\fR \f(CW\*(C`string\*(C'\fR Execute system command
.PP
Command \fB:e\fR executes a system command. 
.PP
\&\fB:q\fR           Quit
.PP
This command quits an interactive pknotsRG session.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
Start pknotsRG in interactive mode:
.PP
.Vb 1
\&    pknotsRG
.Ve
.PP
Pipe file input.seq into the program:
.PP
.Vb 1
\&    cat input.seq | pknotsRG
.Ve
.SH "EXIT STATUS"
.IX Header "EXIT STATUS"
pknotsRG returns a zero exit status if it succeeds. Non zero is
returned in case of failure.
.SH "AUTHORS"
.IX Header "AUTHORS"
Copyright (C) 2004\-2006 Jens Reeder, Robert Giegerich, Peter Steffen
.PP
Faculty of Technology, Bielefeld University, 33594 Bielefeld, Germany
.PP
Send comments/bug reports to: J.Reeder <jreeder@techfak.uni\-bielefeld.de>.
Updates: http://bibiserv.techfak.uni\-bielefeld.de/pknotsrg/
.PP
\&\fBpknotsRG contains code from the following contributors:\fR
.PP
\&\fBEditline library (libedit)\fR
.PP
Copyright (C) 1997\-2003 The NetBSD Foundation, Inc.
All rights reserved.


pknotsRG-1.3/ChangeLog
Version 1.3: September 2006
        * no need for garbage collector library anymore
	* interactive mode
	* support for fasta file input
	* new option -w: sliding window
	* new option -S: output width for structures
	
Version 1.2: February 2005
	* Faster suboptimal modes (at least factor 10)

Version 1.1: 
	* Minor Changes
        * New option -o: no suboptimals inside pseudoknots

	
Version 1.0: 
	* Initial version


pknotsRG-1.3/src/pknotsRG.c
/* ---------------------------------------------------------------------------
pknotsRG 1.3
Copyright (C) 2004-2006 Jens Reeder, Robert Giegerich, Peter Steffen
--------------------------------------------------------------------------- */

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include "config.h"
#include "options.h"
#include "adplib.h"
#include "pknotsRG.h"

/* ====================================================================== 
   Reset mode 
   ====================================================================== */

static void reset_mode(toptions *opt) {
  opt->colored_output = 1; // Colored output (-z)
}


/* ====================================================================== 
   Init default values
   ====================================================================== */

static void init_defaults(toptions *opt) {
  opt->traceback_diff    = 0;    // Set energy range (kcal/mol) (-e)
  opt->traceback_percent = 10;   // Set energy range (%%) (-c)
  opt->window_mode       = 0;    // Set window size (-w)
  opt->window_size       = 0;    // Set window size (-w)
  opt->window_size_user  = 0;    // Set window size (-w)
  opt->window_step       = 1;    // Set window position increment (-W)
  opt->split_output_mode = 0;    // Specify output width for structures (-S)
  opt->split_output_size = 0;    // Specify output width for structures (-S)
  if (opt->inputfile) free(opt->inputfile);
  opt->inputfile         = NULL; // Read input from file (-f)
  opt->colored_output    = 0;    // Colored output (-z)
  opt->terminate            = 0;
  opt->interactive          = 0;
  /* output mode */
  opt->output_mode          = 0;
  opt->format_string        =NULL;

  opt->debug=0;

  opt->alg = 0;
  opt->so = 0;
  opt->nd = 0;
  opt->pksOnly = 0;

  wkn              = 1;    /* The weighting parameter for pseudoknots */
  npp              = 30;   /* Unpaired base in a pseudoknot */
  pbp              = 0;    /* Basepair in a pseudoknot */
  pkinit           = 900;  /* pseudoknot - init - penalty */
  max_pknot_length      = -1;   /* maximal pseudoknot-length (-1 -> maximal) */
  max_pknot_length_user = -1;   /* maximal pseudoknot-length (-1 -> maximal) */
}



/* ====================================================================== 
   Process arguments   
   ====================================================================== */

static void process_args(toptions *opt, char interactive, int argc, char **argv) {
  int   c;
  char  manopt, manoptmode;
  float temp;

  opterr = 0; optind = 1;

  pcolor(opt->colored_output,COLOR_BLUE);
  while ((c = getopt (argc, argv, "hH:ve:c:w:W:S:F:zmflsudon:p:k:")) != -1) {
    switch (c) {

      // Display this information (-h)
      case 'h':
        if (interactive) {
          printf("Interactive mode:\nEnter sequence directly or use the following commands to change settings:\n");
        }
        else {
        
        printf("Options:\n");
        }
        printf("  -h           	Display this information\n");
        printf("  -H <option>  	Display detailed information on <option>\n");
        printf("  -v           	Show version\n");
        printf("  -m            Use mfe strategy\n");
        printf("  -f            Use enf strategy\n");
        printf("  -l            Use loc strategy\n");
        printf("  -s            Show suboptimals\n");
        printf("  -u            no dangling bases (implies -s)\n");
        printf("  -o            no suboptimals inside pknots (implies -s -l)\n");
        printf("  -e <value>   	Set energy range (kcal/mol)\n");
        printf("  -c <value>   	Set energy range (%%)\n");
        printf("  -n <value>    Set npp-value [0.3]\n");
        printf("  -p <value>    Set pkinit-value [9]\n");
        printf("  -k <value>    Set maximal pknot-length\n");
        printf("  -w <value>   	Set window size\n");
        printf("  -W <value>   	Set window position increment\n");
        printf("  -S <value>   	Specify output width for structures\n");
        printf("  -F <filename>	Read input from file\n");
        printf("  -z           	Colored output\n");
        if (opt->interactive) {
        printf("Additional interactive mode commands:\n");
        printf("  :s           	Show current configuration\n");
        printf("  :d           	Reset configuration\n");
        printf("  :e <string>  	Execute system command\n");
        printf("  :q           	Quit\n");
        }
        opt->terminate = 1;
        break;
      // Display detailed information on <option> (-H)
      case 'H':
        if      (optarg[0]=='-') { manoptmode = '-'; manopt = optarg[1]; }
        else if (optarg[0]==':') { manoptmode = ':'; manopt = optarg[1]; }
        else                     { manoptmode = '-'; manopt = optarg[0]; }
        if (!interactive) printf("\n");
        #include "pknotsRG-man.c"
        if (!interactive) printf("\n");
        opt->terminate = 1;
        break;
      // Show version (-v)
      case 'v':
        printf("%s (%s)\n",PACKAGE_STRING,RELEASE_DATE);
        printf("Copyright (C) 2004-2006 Jens Reeder, Robert Giegerich, Peter Steffen\n");
        printf("Send comments/bug reports to: J.Reeder <jreeder@techfak.uni-bielefeld.de>.\n");
        printf("Updates: http://bibiserv.techfak.uni-bielefeld.de/pknotsrg/\n");
        printf("\n");
        printf("pknotsRG contains code from the following contributors:\n");
        printf("\n");
        printf("* Editline library (libedit):\n");
        printf("  Copyright (C) 1997-2003 The NetBSD Foundation, Inc.\n");
        printf("  All rights reserved.\n");
        opt->terminate = 1;
        break;
      // Use mfe strategy
      case 'm':
        opt->alg = 0;
        if (interactive) printf("Use mfe strategy.\n");
        break;
      // Use enf strategy
      case 'f':
        opt->alg = 1;
        if (interactive) printf("Use enf strategy.\n");
        break;
      // Use loc strategy
      case 'l':
        opt->alg = 2;
        if (interactive) printf("Use loc strategy.\n");
        break;
      // Show suboptimals
      case 's':
        if (interactive) {
	  printf("Show suboptimals");
	  opt->so = 1 - opt->so;
	  if (opt->so) printf(" = ON. Type -s again to switch off.\n");
	  else printf(" = OFF. Type -s again to switch on.\n");
	}
        else opt->so = 1;
        break;
      // No dangling bases
      case 'u':
        if (interactive) {
	  printf("No dangling bases");
	  opt->nd = 1 - opt->nd;
	  if (opt->nd) opt->so=1;
	  if (opt->nd) printf(" = ON. Type -u again to switch off.\n");
	  else printf(" = OFF. Type -u again to switch on.\n");
	}
	else {
          opt->nd=1;
          opt->so=1;
	}
        break;
      // No suboptimals inside pknots
      case 'o':
        if (interactive) {
	  printf("No suboptimals inside pknots");
	  opt->pksOnly = 1 - opt->pksOnly;
	  if (opt->pksOnly) {opt->so=1; opt->alg=2;}
	  if (opt->pksOnly) printf(" = ON. Type -o again to switch off.\n");
	  else printf(" = OFF. Type -o again to switch on.\n");
	}
	else {
          opt->pksOnly=1;
          opt->so=1;
          opt->alg=2;
	}
        break;
      // Debug mode
      case 'd':
        opt->debug=1;
	break;
      // Set energy range (kcal/mol) (-e)
      case 'e':
        sscanf(optarg,"%f",&(opt->traceback_diff));
        opt->traceback_diff = max(0, opt->traceback_diff);
        if (interactive) printf("Energy range set to %.2f kcal/mol.\n", opt->traceback_diff);
        
        opt->traceback_percent = 0;
        break;
      // Set energy range (%%) (-c)
      case 'c':
        sscanf(optarg,"%d",&(opt->traceback_percent));
        opt->traceback_percent = max(0, opt->traceback_percent);
        if (interactive) printf("Energy range set to %d%% of mfe.\n", opt->traceback_percent);
        
        opt->traceback_diff = 0;
        break;
      // Set npp value
      case 'n':
        sscanf(optarg,"%f",&temp);
	if (interactive) printf("npp-value set to %.2f.\n", temp);
        npp = (int) (temp * 100);
        break;
      // Set pkinit value
      case 'p':
        sscanf(optarg,"%f",&temp);
	if (interactive) printf("pkinit-value set to %.2f.\n", temp);
        pkinit = (int) (temp * 100);
        break;
      // Set maximal pknot-length
      case 'k':
        sscanf(optarg,"%d",&max_pknot_length_user);
	if (interactive) printf("Maximal pknot-length set to %d.\n", max_pknot_length_user);
        break;
      // Set window size (-w)
      case 'w':
        if ((interactive) && (optarg[0] == '-')) {
          printf("Window mode disabled.\n");
          opt->window_mode = 0;
        }
        if (optarg[0] != '-') {
          opt->window_mode = 1;
          sscanf(optarg,"%d",&(opt->window_size_user));
          opt->window_size_user = max(1, opt->window_size_user);
          if (interactive) printf("Set window size to %d. Type -w - to disable.\n", opt->window_size_user);
        }
        break;
      // Set window position increment (-W)
      case 'W':
        sscanf(optarg,"%d",&(opt->window_step));
        opt->window_step = max(1, opt->window_step);
        if (interactive) printf("Set window position increment to %d.\n", opt->window_step);
        break;
      // Specify output width for structures (-S)
      case 'S':
        if ((interactive) && (optarg[0] == '-')) {
          printf("Output splitting disabled.\n");
          opt->split_output_mode = 0;
        }
        if (optarg[0] != '-') {
          opt->split_output_mode = 1;
          sscanf(optarg,"%d",&(opt->split_output_size));
          opt->split_output_size = max(1, opt->split_output_size);
          if (interactive) printf("Set output size to %d. Type -S - to disable.\n", opt->split_output_size);
        }
        break;
      // Read input from file (-F)
      case 'F':
        if (opt->inputfile) free(opt->inputfile);
        opt->inputfile = mkstr(optarg);
        break;
      // Colored output (-z)
      case 'z':
        if (interactive) {
          printf("Colored output");
          opt->colored_output = 1 - opt->colored_output;
          if (opt->colored_output) printf (" = ON. Type -z again to switch off.\n");
          else printf (" = OFF. Type -z again to switch on.\n");
        }
        else opt->colored_output = 1;
        if (opt->interactive && !opt->colored_output) printf(COLOR_DEFAULT);
        break;
      case '?':
        if (interactive) printf("Unknown option '-%c'. Type -h for more information.\n", optopt);
        else {
          fprintf (stderr, "%s: unknown option `-%c'.\nTry '%s -h' for more information.\n", argv[0], optopt, argv[0]);
          exit(1);
        }
        break;
      default:
        abort ();
      }
  }
  pcolor(opt->colored_output,COLOR_DEFAULT);
}

/* ====================================================================== 
   Print settings
   ====================================================================== */

static void print_settings(toptions *opt) {
 
  printf("Current settings\n-----------------\n");
  printf("Current strategy              ");
  if (opt->alg==0) printf("mfe\n");
  else if (opt->alg==1) printf("enf\n");
  else if (opt->alg==2) printf("loc\n");
  printf("Show suboptimals              %s  (-s)\n", opt->so ? "ON" : "OFF");
  printf("No dangling bases             %s  (-u)\n", opt->nd ? "ON" : "OFF");
  printf("No suboptimals inside pknots  %s  (-o)\n", opt->pksOnly ? "ON" : "OFF");
  printf("Energy range                  %.2f kcal/mol  (-e)\n", opt->traceback_diff);
  printf("Energy range                  %d%% of mfe  (-c)\n", opt->traceback_percent);
  printf("npp-value                     %.2f  (-n)\n", (float) npp / 100);
  printf("pkinit-value                  %.2f  (-p)\n", (float) pkinit / 100);
  if (max_pknot_length_user == -1) 
  printf("Maximal pknot-length          <input length>  (-k)\n");
  else 
  printf("Maximal pknot-length          %d  (-k)\n", max_pknot_length_user);
  if (opt->window_mode)
  printf("Window size:                  %d  (-w)\n", opt->window_size_user);
  else
  printf("Window size:                  %s-  (-w)\n","");
  printf("Window position increment:    %d  (-W)\n", opt->window_step);
  if (opt->split_output_mode)
  printf("Output width for structures:  %d  (-S)\n", opt->split_output_size);
  else
  printf("Output width for structures:  %s-  (-S)\n","");
  printf("Colored output                %s  (-z)\n", opt->colored_output ? "ON" : "OFF" );
}


/* ====================================================================== 
   Main interface
   ====================================================================== */

int main (int argc, char **argv) {
  toptions   *opt;
  tsequence  *seq;
  treadseq   *rs        = NULL; 
  ttokenizer *tokenizer = NULL;
  char       *command;

  opt = (toptions *) calloc(1,sizeof(toptions));

  init_defaults(opt);
  process_args(opt, 0, argc, argv);

  if (!opt->terminate) {
    if (optind < argc)               rs = readseq_open(READSEQ_STRING, argv[optind]);
    else if (opt->inputfile)         rs = readseq_open(READSEQ_FILE,   opt->inputfile);
    else if (!isatty(fileno(stdin))) rs = readseq_open(READSEQ_STDIN,  NULL);
    else {
      printf("Interactive mode. Try `./pknotsRG -h` for more information.\n", argv[0]);
      rl_init();
      opt->interactive = 1;
      opt->colored_output = 1 - opt->colored_output;
      tokenizer = tokenizer_new();
      rs = readseq_open(READSEQ_STRING, "");
    } 

    while (1) {
      if (opt->interactive) {
        if (opt->colored_output) 
          printf("%s\nInput sequence (upper or lower case); :q to quit, -h for help.\n....,....1....,....2....,....3....,....4....,....5....,....6....,....7....,....8\n%s",COLOR_RED,COLOR_DEFAULT);
        else 
	  printf("\nInput sequence (upper or lower case); :q to quit, -h for help.\n....,....1....,....2....,....3....,....4....,....5....,....6....,....7....,....8\n");
        command = rl_gets();

        if (!command || (command[0] == '@') || ((command[0] == ':') && (command[1] == 'q'))) {
          pcolor(opt->colored_output,COLOR_BLUE);
          printf("Leaving pknotsRG.");
          pcolor(opt->colored_output,COLOR_DEFAULT);
          printf("\n");
          exit(0);
        }
        else if (command[0] == ':') {
          pcolor(opt->colored_output,COLOR_BLUE);
          if (command[1] == 's') print_settings(opt);
          if (command[1] == 'd') {
            init_defaults(opt);
            opt->colored_output = 1;
            opt->interactive = 1;
            printf("Activated default configuration.\n");
            pcolor(opt->colored_output,COLOR_DEFAULT);
          }
          if (command[1] == 'e') {
            system(command + 2);
          }
        }
        else if (command[0] == '-') {
          tokenizer_exec(tokenizer, argv[0], command);
          process_args(opt, 1, tokenizer->count, tokenizer->token);
          if (opt->inputfile) { 
            rs = readseq_free(rs);
            rs = readseq_open(READSEQ_FILE, opt->inputfile);
          }
          free(opt->inputfile);
          opt->inputfile = NULL;
        }
        else {
          rs = readseq_free(rs);
          rs = readseq_open(READSEQ_STRING, command);
        }
      }

      while (1) {
        seq = readseq_next_fasta(rs);
        if (!(seq->success)) break;
        if (max_pknot_length_user == (-1)) max_pknot_length=seq->length;
        else                               max_pknot_length=max_pknot_length_user;

        if (!opt->so) {
          if (opt->alg==0) main_pk_mfe(opt,seq);
          if (opt->alg==1) main_pk_enf(opt,seq);
          if (opt->alg==2) main_pk_loc(opt,seq);
        } else {
          if (!opt->nd) {
            if (opt->alg==0) main_pk_mfe_so(opt,seq);
            if (opt->alg==1) main_pk_enf_so(opt,seq);
            if (opt->alg==2) main_pk_loc_so(opt,seq);
          }
          else {
            if (opt->alg==0) main_pk_mfe_so_nd(opt,seq);
            if (opt->alg==1) main_pk_enf_so_nd(opt,seq);
            if (opt->alg==2) main_pk_loc_so_nd(opt,seq);
          }
        }
        sequence_free(seq);
      }

      if (!opt->interactive) break;
    }
  }
  exit(0);
}
